conda_recipe_manager/types.py

```diff
@@ -35,9 +35,22 @@
 H = TypeVar("H", bound=Hashable)
 
 
-# All sentinel values used in this module should be constructed with this class, for typing purposes.
 class SentinelType:
-    pass
+    """
+    A single sentinel class to be used in this project, as an alternative to `None` when `None` cannot be used.
+    It is defined in a way such that SentinelType instances survive pickling and allocations in different memory
+    spaces.
+    """
+
+    def __eq__(self, o: object) -> bool:
+        """
+        Provides a pickle/thread-safe way to compare objects to the Sentinel Type. All sentinels are equal by simply
+        being sentinel objects.
+        :param o: The object to compare against.
+        :returns: True if the object is a SentinelType object.
+        """
+        # There is an implicit `isinstance(self, SentinelType) and ...` to this logic
+        return isinstance(o, SentinelType)
```

Comment:
I'd suggest reusing some of the sentinel implementations that exist out there, unittest.mock has one, there is https://github.com/Infinidat/sentinels or https://github.com/eddieantonio/sentinel and probably many more.

I don't think your implementation covers the full list of qualities of a sentinel value (e.g. `_sentinel` isn't able to be compared with `is`), even though it might fit this particular PR's needs. Nothing wrong with sentinels generally, of course, they are a great pattern.

Reply:
I've struggled to find a standard approach here. I would prefer to use something from the standard library, but the mocking version looks specifically designed for testing. I don't think I want unit test helpers in production code.

The other two libraries listed look like they haven't had a release in a few years and I'm unsure what level of support we can expect.

`is` is not used in any of the sentinel checks that currently exist. For our purposes, the current solution may be good enough for now. We need _some_ kind of sentinel for the `RecipeParser`, there are a number of places I need to discern the difference between `None`/`Null` and "another default of northing".

Reply:
It'd be nice if this was accepted: https://peps.python.org/pep-0661/

Reply:
> It'd be nice if this was accepted: [peps.python.org/pep-0661](https://peps.python.org/pep-0661/)

True!

FWIW, I was suggesting to simply copy over the code, not rely on it as a dependency. They are 3-clause BSD and MIT licensed, so are pretty simple to include, as long as the license statement is kept.

Reply:
That said, it was really just a suggestion, not a suggestion wink wink :) I'll have more time to finish the review tomorrow!